scripts/scr_powers_new/scr_powers_new.gml

```diff
@@ -1,56 +1,57 @@
-function scr_powers_new(argument0, argument1) {
+function scr_powers_new(){
 
-	// Argument0: Company
-	// Argument1: Marine ID
+	// company: Company
+	// marine_number: Marine ID
 
 	// This script checks if the marine is capable of using psychic powers, and if so, assigns them based on experience
 
 	var random_learn;
 	random_learn=false;
 
-	var let,letmax,powers_should_have,powers_have;
-	let="";letmax=0;
+	var powers_should_have,powers_have;
+	var power_code="";letmax=0;
 
-	if (obj_ini.psy_powers="default"){let="D";letmax=7;}
-	if (obj_ini.psy_powers="biomancy"){let="B";letmax=5;}
-	if (obj_ini.psy_powers="pyromancy"){let="P";letmax=5;}
-	if (obj_ini.psy_powers="telekinesis"){let="T";letmax=5;}
-	if (obj_ini.psy_powers="rune Magick") or (obj_ini.psy_powers="Rune Magick"){let="R";letmax=5;}
+	if (obj_ini.psy_powers="default"){power_code="D";letmax=7;}
+	if (obj_ini.psy_powers="biomancy"){power_code="B";letmax=5;}
+	if (obj_ini.psy_powers="pyromancy"){power_code="P";letmax=5;}
+	if (obj_ini.psy_powers="telekinesis"){power_code="T";letmax=5;}
+	if (obj_ini.psy_powers="rune Magick") or (obj_ini.psy_powers="Rune Magick"){power_code="R";letmax=5;}
 
 
-
-	powers_should_have=floor((obj_ini.experience[argument0,argument1]-30)/30)+1;// +1 for the primary
-	powers_have=string_count(string(let),obj_ini.spe[argument0,argument1]);
+	// higer psionice means more powers learnt
+	powers_should_have=floor((experience()-30)/(45-psionic))+1;// +1 for the primary
+	powers_have=string_count(string(power_code),specials());
 
 
 
 
 	if (powers_have<powers_should_have) and (powers_have<(letmax)+1) and (random_learn=true){
 	    var newpow;newpow=0;
-	    repeat(100){if (powers_have<powers_should_have)and (powers_have<(letmax)+1){
-	        var tha;tha=floor(random(letmax))+1;
-	        if (string_count(string(tha),obj_ini.spe[argument0,argument1])=0){
-	            powers_have+=1;obj_ini.spe[argument0,argument1]+=string(let)+string(tha)+"|";
-	        }
-	    }}
+	    if (powers_have<powers_should_have) and (powers_have<(letmax)+1){
+		    if (powers_have<powers_should_have) and (powers_have<(letmax)+1){
+		        var tha=floor(random(letmax))+1;
+		        if (string_count(string(tha),specials())=0){
+		            powers_have+=1;
+		            obj_ini.spe[company,marine_number]+=string(power_code)+string(tha)+"|";
+		        }
+		    }
+		}
 	}
 
 
 	if (powers_have<powers_should_have) and (powers_have<(letmax)+1) and (random_learn=false){
```

Comment:
Should note that this means starting at line 42 since this path is to account for not randomly learning, but the above suggestions merge the logic into a single `while` loop